File: mjx/hand.cpp
#include "mjx/hand.h"

#include "mjx/internal/shanten_calculator.h"

#include <google/protobuf/util/json_util.h>
#include <google/protobuf/util/message_differencer.h>

#include <utility>

namespace mjx {
Hand::Hand(mjxproto::Hand proto) : proto_(std::move(proto)) {}

Hand::Hand(const std::string& json) {
  auto status = google::protobuf::util::JsonStringToMessage(json, &proto_);
  assert(status.ok());
}

const mjxproto::Hand& mjx::Hand::ToProto() const noexcept { return proto_; }

std::string mjx::Hand::ToJson() const noexcept {
  std::string serialized;
  auto status =
      google::protobuf::util::MessageToJsonString(proto_, &serialized);
  assert(status.ok());
  return serialized;
}

bool Hand::operator==(const Hand& other) const noexcept {
  return google::protobuf::util::MessageDifferencer::Equals(proto_,
                                                            other.proto_);
}

bool Hand::operator!=(const Hand& other) const noexcept {
  return !(*this == other);
}

std::array<uint8_t, 34> Hand::ClosedTiles() const noexcept {
  std::array<uint8_t, 34> closed_tiles{};
  closed_tiles.fill(0);
  for (auto t : proto_.closed_tiles()) {
    ++closed_tiles[t];
  }
  return closed_tiles;
}

bool Hand::IsTenpai() const {
  return mjx::internal::ShantenCalculator::ShantenNumber(ClosedTiles()) == 1;
}

int Hand::ShantenNumber() const {
  return mjx::internal::ShantenCalculator::ShantenNumber(ClosedTiles());
}

}  // namespace mjx

File: mjx/hand.py
from __future__ import annotations

from typing import List, Optional

import _mjx  # type: ignore
from google.protobuf import json_format

import mjxproto
from mjx.const import TileType
from mjx.open import Open
from mjx.tile import Tile


class Hand:
    def __init__(self, hand_json: Optional[str] = None) -> None:
        self._cpp_obj: Optional[_mjx.Hand] = None  # type: ignore
        if hand_json is None:
            return

        self._cpp_obj = _mjx.Hand(hand_json)  # type: ignore

    def closed_tiles(self) -> List[Tile]:
        assert self._cpp_obj is not None  # type: ignore
        return [Tile(tid) for tid in self._cpp_obj.closed_tiles()]  # type: ignore

    def closed_tile_types(self) -> List[TileType]:
        assert self._cpp_obj is not None  # type: ignore
        return [TileType(tt) for tt in self._cpp_obj.closed_tile_types()]  # type: ignore

    def opens(self) -> List[Open]:
        assert self._cpp_obj is not None  # type: ignore
        return [Open(bit) for bit in self._cpp_obj.opens()]  # type: ignore

    def shanten_number(self) -> int:
        assert self._cpp_obj is not None  # type: ignore
        return self._cpp_obj.shanten_number()  # type: ignore

    def effective_draw_types(self) -> List[TileType]:
        "Retrun the list of tile types which reduce the shanten number by drawing it"
        return [TileType(i) for i in self._cpp_obj.effective_draw_types()]  # type: ignore

    def effective_discard_types(self) -> List[TileType]:
        "Retrun the list of tile types which reduce the shanten number by discarding it"
        return [TileType(i) for i in self._cpp_obj.effective_discard_types()]  # type: ignore

    def to_json(self) -> str:
        assert self._cpp_obj is not None  # type: ignore
        return self._cpp_obj.to_json()  # type: ignore

    def to_proto(self) -> mjxproto.Hand:
        assert self._cpp_obj is not None  # type: ignore
        return json_format.Parse(self.to_json(), mjxproto.Hand())

    @classmethod
    def _from_cpp_obj(cls, cpp_obj: _mjx.Hand) -> Hand:  # type: ignore
        hand = cls()
        hand._cpp_obj = cpp_obj  # type: ignore
        return hand
